File: platforms/bukkit/runtime/src/main/java/ru/privatenull/pnlibrary/remote/bukkit/RemoteClassLoader.java
package ru.privatenull.pnlibrary.remote.bukkit;

import java.io.Closeable;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;

final class RemoteClassLoader implements Closeable {
    private final ClassLoader loader;
    private final java.nio.file.Path temporaryRoot;

    private RemoteClassLoader(ClassLoader loader) { this(loader, null); }
    private RemoteClassLoader(ClassLoader loader, java.nio.file.Path temporaryRoot) { this.loader = loader; this.temporaryRoot = temporaryRoot; }
    static RemoteClassLoader fromCompiled(ClassLoader loader, java.nio.file.Path root, String className) { return new RemoteClassLoader(new NamedLoader(loader, className), root); }

    static RemoteClassLoader forBytes(byte[] bytes, String className, ClassLoader parent) throws Exception {
        if (bytes.length < 8 || bytes[0] != 'P' || bytes[1] != 'K') {
            verifyBytecodeLevel(bytes);
            return new RemoteClassLoader(new SingleClassLoader(bytes, className, parent));
        }
        java.nio.file.Path jar = java.nio.file.Files.createTempFile("pnlibrary-remote-class-", ".jar");
        java.nio.file.Files.write(jar, bytes);
        return new RemoteClassLoader(new URLClassLoader(new URL[] { jar.toUri().toURL() }, parent) {
            @Override public void close() throws IOException { super.close(); java.nio.file.Files.deleteIfExists(jar); }
        });
    }

    Class<?> load(String name) throws ClassNotFoundException { return Class.forName(name, true, loader); }
    @Override public void close() throws IOException { if (loader instanceof Closeable) ((Closeable) loader).close(); if (temporaryRoot != null) delete(temporaryRoot); }
    private static void delete(java.nio.file.Path root) throws IOException { if (Files.exists(root)) Files.walk(root).sorted(java.util.Comparator.reverseOrder()).forEach(path -> { try { Files.deleteIfExists(path); } catch (IOException ignored) { } }); }

    private static void verifyBytecodeLevel(byte[] bytes) {
        if (bytes.length < 8 || bytes[0] != (byte) 0xCA || bytes[1] != (byte) 0xFE || bytes[2] != (byte) 0xBA || bytes[3] != (byte) 0xBE) {
            throw new IllegalArgumentException("remote file is neither a .class file nor a JAR");
        }
        int major = ((bytes[6] & 0xff) << 8) | (bytes[7] & 0xff);
        String specification = System.getProperty("java.specification.version", "8");
        int java = specification.startsWith("1.") ? Integer.parseInt(specification.substring(2)) : Integer.parseInt(specification.split("\\.")[0]);
        int supported = 44 + java;
        if (major > supported) throw new IllegalArgumentException("remote class requires Java bytecode " + major + ", server supports " + supported + " (Java " + java + ")");
    }

    private static final class SingleClassLoader extends ClassLoader {
        private final byte[] bytes; private final String name;
        SingleClassLoader(byte[] bytes, String name, ClassLoader parent) { super(parent); this.bytes = bytes; this.name = name; }
        @Override protected Class<?> findClass(String requested) throws ClassNotFoundException {
            if (!name.equals(requested)) throw new ClassNotFoundException(requested);
            return defineClass(requested, bytes, 0, bytes.length);
        }
    }
    private static final class NamedLoader extends ClassLoader {
        private final ClassLoader delegate; private final String name;
        NamedLoader(ClassLoader delegate, String name) { super(delegate.getParent()); this.delegate = delegate; this.name = name; }
        @Override protected Class<?> findClass(String requested) throws ClassNotFoundException { if (!name.equals(requested)) throw new ClassNotFoundException(requested); return delegate.loadClass(requested); }
    }
}

File: platforms/bukkit/runtime/src/main/kotlin/ru/privatenull/pnlibrary/remote/bukkit/RemoteClassLoader.kt
package ru.privatenull.pnlibrary.remote.bukkit

import java.io.Closeable
import java.net.URLClassLoader
import java.nio.file.Files
import java.nio.file.Path

internal class RemoteClassLoader private constructor(
    private val loader: ClassLoader,
    val className: String,
    private val temporaryRoot: Path? = null,
) : Closeable {
    fun load(): Class<*> = Class.forName(className, true, loader)

    override fun close() {
        (loader as? Closeable)?.close()
        temporaryRoot?.let { root ->
            if (Files.exists(root)) Files.walk(root).use { paths ->
                paths.sorted(Comparator.reverseOrder()).forEach { Files.deleteIfExists(it) }
            }
        }
    }

    companion object {
        fun compiled(root: Path, name: String, parent: ClassLoader): RemoteClassLoader =
            RemoteClassLoader(URLClassLoader(arrayOf(root.toUri().toURL()), parent), name, root)

        fun forBytes(bytes: ByteArray, name: String?, parent: ClassLoader): RemoteClassLoader {
            require(!name.isNullOrBlank()) { "class name is required for compiled policy" }
            if (bytes.size >= 2 && bytes[0] == 'P'.code.toByte() && bytes[1] == 'K'.code.toByte()) {
                val root = Files.createTempDirectory("pnlibrary-remote-jar-")
                val jar = root.resolve("policy.jar")
                Files.write(jar, bytes)
                return RemoteClassLoader(URLClassLoader(arrayOf(jar.toUri().toURL()), parent), name, root)
            }
            require(bytes.size >= 8 && bytes[0] == 0xCA.toByte() && bytes[1] == 0xFE.toByte() && bytes[2] == 0xBA.toByte() && bytes[3] == 0xBE.toByte()) {
                "remote file is not a Java class"
            }
            val major = ((bytes[6].toInt() and 255) shl 8) or (bytes[7].toInt() and 255)
            val java = System.getProperty("java.specification.version", "8").removePrefix("1.").substringBefore('.').toInt()
            require(major <= 44 + java) { "remote class requires newer Java" }
            val loader = object : ClassLoader(parent) {
                override fun findClass(requested: String): Class<*> {
                    if (requested != name) throw ClassNotFoundException(requested)
                    return defineClass(requested, bytes, 0, bytes.size)
                }
            }
            return RemoteClassLoader(loader, name)
        }
    }
}
